includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -559,6 +559,15 @@ private function process_node( DOMElement $node ) {
 			}
 		}
 
+		// Remove the element if it has a invalid layout attributes.
+		if (
+			! $this->is_valid_layout( $tag_spec, $node ) &&
+			$this->remove_invalid_child( $node, [ 'code' => 'invalid_layout' ] )
+		) {
+			$this->remove_node( $node );
+			return null;
+		}
+
```

Comment:
There is a merge conflict now with this code:

![image](https://user-images.githubusercontent.com/134745/70676845-d7f2a480-1cd0-11ea-9735-402e0bb85ca1.png)

Now instead of using `invalid_layout` string, a new error code constant should be introduced. For example:

```diff
--- a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
+++ b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
@@ -55,6 +55,7 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
 	const INVALID_BLACKLISTED_VALUE_REGEX      = 'INVALID_BLACKLISTED_VALUE_REGEX';
 	const DISALLOWED_PROPERTY_IN_ATTR_VALUE    = 'DISALLOWED_PROPERTY_IN_ATTR_VALUE';
 	const ATTR_REQUIRED_BUT_MISSING            = 'ATTR_REQUIRED_BUT_MISSING';
+	const INVALID_LAYOUT                       = 'INVALID_LAYOUT';
 
 	/**
 	 * Allowed tags.
```

I think it can be resolved as:

```php

		// Remove the element if it has a invalid layout attributes.
		if (
			! $this->is_valid_layout( $tag_spec, $node ) &&
			$this->remove_invalid_child( $node, [ 'code' => self::INVALID_LAYOUT ] )
		) {
			$this->remove_node( $node );
			return null;
		}

		// Identify attribute values that don't conform to the attr_spec.
		$disallowed_attributes = $this->sanitize_disallowed_attribute_values_in_node( $node, $merged_attr_spec_list );
```

Reply:
Actually instead of there being one `INVALID_LAYOUT` error, the `is_valid_layout` method could now return not just a boolean but rather `true` when success and an error array when a failure happens, which will indicate the specific error that is happening. So for example:

```diff
diff --git a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
index dd608d58..7deb7935 100644
--- a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
+++ b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
@@ -1142,8 +1142,11 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
 		$input_width->validate( $allow_auto, $allow_fluid );
 		$input_height->validate( $allow_auto, $allow_fluid );
 
-		if ( ! $input_width->is_valid() || ! $input_height->is_valid() ) {
-			return false;
+		if ( ! $input_width->is_valid() ) {
+			return [ 'code' => self::INVALID_LAYOUT_WIDTH ];
+		}
+		if ( ! $input_height->is_valid() ) {
+			return [ 'code' => self::INVALID_LAYOUT_HEIGHT ];
 		}
 
 		// No need to go further if there is no layout attribute.
@@ -1161,7 +1164,7 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
 
 		// Only FLEX_ITEM allows for height to be set to auto.
 		if ( $height->is_auto() && AMP_Rule_Spec::LAYOUT_FLEX_ITEM !== $layout ) {
-			return false;
+			return [ 'code' => self::INVALID_LAYOUT_AUTO_HEIGHT ];
 		}
 
 		// FIXED, FIXED_HEIGHT, INTRINSIC, RESPONSIVE must have height set.
```

And so on. 

The conditional to remove the element can then be changed to:

```php
// Remove the element if it has a invalid layout attributes.
$layout_validity = $this->is_valid_layout( $tag_spec, $node )
if ( true !== $layout_validity && $this->remove_invalid_child( $node, $layout_validity ) ) {
	return null;
}
```

Reply:
That was a fun merge conflict :sweat_smile:. I've added error codes for each invalid layout in https://github.com/ampproject/amp-wp/pull/3728/commits/6a6ab8ce3d509d37d08d6ee8ecb2beb208ac8866,